I2C_eeprom_cyclic_store.h

```diff
@@ -0,0 +1,231 @@
+// PURPOSE: Supplemental utility class for I2C_EEPROM library
+//
+
+
+#include <I2C_eeprom.h>
+
+/**
+ * @brief This is a utility class for using an eeprom to store a simple
+ * data structure.
+ * 
+ * The purpose of the utility is to extend the life of an eeprom memory by
+ * rotating the writes over different pages to reduce the number of writes
+ * to any individual page.
+ * It does this by partitioning the memory into slots (of continuous pages)
+ * big enough to contain the data structure. When data is written to a slot
+ * it is given a header with a
+ * version number.
+ * On initialization the slots are scanned for the one with the highest
+ * version number which is then used for subsequent reads.
+ * Whenever data is written the version number is incremented and the next
+ * slot in sequence is used (or the first slot if going past the end).
+ *  
+ * Note that the data is stored in binary form which means it should not be
+ * expected that the eeprom can be moved between architectures. Data stored
+ * in an eeprom will also become invalid if the data structure is changed
+ * either manually or due to changed optimization settings.
+ *
+ * If the data structure has changed or if the eeprom contains other data it
+ * must first be formatted with a call to format().
+ * 
+ * Finally, since the version number is a long word, this class will
+ * *derail and fail* to add new versions past 4294967295 writes.
+ * 
+ * @tparam T the type of the data structure to store, should only contain
+ * **value** members and no constructor/destructor nor other
+ * methods/functions - e g a pure DTO.
+ */
+template<typename T>
+class I2C_eeprom_cyclic_store
+{
+    public:
+        /**
+        * Initializes the instance with the I2C_eeprom object to use
+        * This call searches the eeprom for the latest written version and
+	* sets the current slot accordingly.
+        * 
+        * @param eeprom  the instance of I2C_eeprom to use
+        * @param pageSize is the number of bytes in each write page
+        * @param totalPages specifies the total number of pages to use. By
+	* specifying a number that i sless than the available pages will
+	* exclude the remaining pages from being used.
+        * @return true if initialization succeeds, false otherwise
+        */
+        bool begin(I2C_eeprom & eeprom, uint8_t pageSize, uint16_t totalPages)
+        {
+            _eeprom = &eeprom;
+            _pageSize = pageSize;
+            _totalPages = totalPages;
+            auto bufferSize = sizeof(uint32_t)+sizeof(T);
+            _bufferPages = bufferSize/_pageSize+(bufferSize%_pageSize ? 1 : 0);
+
+            return (_bufferPages < _totalPages) && initialize();
+        };
+
+        /**
+         * Formats the eeprom by writing the max version number to each slot 
+         * 
+         * This must be done if the eeprom already contains data or if the
+	 * structure of the data stored changes.
+         */
+        void format()
```

Comment:
Might be faster when using 
```cpp
void format()
{
  _eeprom->setBlock(0, 0xFF, _totalPages * _pageSize);  // this is effectively eeprom size...
  _isEmpty = true;
  _currentSlot = 0;
  _isInitialized = true;
}
```
 you could do a performance test, setBlock uses a larger buffer so it will do less steps.

Thinking about it
 a **format(uint8_t)** function should be in the EEPROM class itself as it just wipes the whole EE.

your format would then become
```cpp
void format()
{
  _eeprom->format(0xFF);
  _isEmpty = true;
  _currentSlot = 0;
  _isInitialized = true;
}
```

Reply:
I agree that it would be neat to erase the entire eeprom, but from what I gather there is no shortcut (e g erase page command) so we'd have to write 0xFF to the entire chip, right? While it internally, in the eeprom, wouldn't make a difference, my approach reduces the I2C traffic to a minimum. Especially when the buffer spans multiple pages.

But this is more or less speculation... 😄 I will make a test.

Reply:
From a few timing runs with a buffer size that is less than the page size, and wth 512 write pages, it takes the old implementation ~1.5 seconds to do a format compared to ~8 seconds using the setBlock() call.